mpremote: add argument --no-soft-reset

jahr opened the pull request:
I added a new comamnd line argument **--no-soft-reset** to mpremote tool. It may be used as the first argument and when used, no soft reset is performed before using exec, eval and run commands. When argument is not used everything works as before.

jahr committed:
mpremote: add argument --no-soft-reset

dpgeorge commented:
Thanks, this is a useful addition.  But I'm thinking if there's a way to make it a command so it's more powerful.

Eg 1) add commands `soft_reset=0` and `soft_reset=1` which set the `soft_reset` flag for subsequent calls to `enter_raw_repl()`.

Eg 2) add commands like `enter_raw_repl_no_soft_reset` (maybe with a shorter name!) and `soft_reset` that give explicit control over entering raw repl and/or doing a soft reset at a given point in the command chain.

jahr commented:
Currently there is a built-in shortcut **reset** which should (if I understand it well) perform immediate soft reset in the command chain, so adding the commend soft-reset seems redundant to me. Other point is that in the current pip version of mpremote (0.0.6) the reset shortcut doesn't work well - using it between two exec commands results in error

`$ mpremote exec "aaa=42" reset exec "print(aaa)"`
`mpremote: 'print(aaa)' is not a command`

However it works well when used as the last command

`$ mpremote --no-soft-reset exec "aaa=42" reset`
`$ mpremote --no-soft-reset exec "print(aaa)"`
`Traceback (most recent call last):`
`File "<stdin>", line 1, in <module>`
`NameError: name 'aaa' isn't defined`

From my point of view fixing the reset shortcut (if it is not fixed already) together with my new argument is the best way. Removing (or leaving and ignoring) the reset shorcut and implementing the soft-reset command would also do the work but it requires much more work (e.g. rewriting the command loop in the main() function, because currently raw_repl is usually entered only once). Point 2) seems too complicated for common use.

dpgeorge commented:
> Currently there is a built-in shortcut **reset** which should (if I understand it well) perform immediate soft reset

Actually, `reset` performs a hard reset of the device, so the connection will be lost after it executes.

Also it's not working in the middle of a command chain because it takes an optional argument, being the timeout to wait before resetting.  And when you do `reset exec` it thinks `exec` is the argument to `reset`.

jahr commented:
> Actually, reset performs a hard reset of the device

I see, I just found it in the code.

So the solution would be adding new commands **no-soft-reset** and **soft-reset** (IMHO better than using soft-reset=0/1) which exits raw repl (if already entered) and changes soft_reset flag value for subsequent exec/eval/run.
I'm afraid that these comamnds will be not easy to understand for many users :-(

jahr committed:
mpremote: instead of the argument --no-soft-reset, new commands soft-reset and no-soft-reset are added

jahr commented:
I've changed it - instead of the argument I added new comamnds. **no-soft-reset** causes subsequent calls of exec/eval/run perform no soft reset and **soft-reset** cancels that behavior, after **soft-reset** next exec/eval/run is run with soft reset before.
It would be good to improve help texts, my English is rather poor ;-)

dpgeorge commented on tools/mpremote/mpremote/main.py:
```diff
@@ -463,9 +468,21 @@ def main():
+            if cmd == "no-soft-reset" and soft_reset:
+                if pyb.in_raw_repl:
+                    pyb.exit_raw_repl()
+                soft_reset = False
```
Why is there a `and soft_reset` in the if-logic?  I think it should run this block of code regardless of whether `soft_reset` is true/false.

Also, it doesn't really need to exit the raw repl.  If it's already in raw repl then it's ok to stay that way.  Removing that would then allow this block to go before the `do_connect()` block above, and so can be used before any connection is made to a device (making it behave more like your original `--no-soft-reset` option).

jahr replied:
> Why is there a `and soft_reset` in the if-logic? I think it should run this block of code regardless of whether `soft_reset` is true/false.

It's because cmd `no-soft-reset` changes how next `enter_raw_repl()` is called and it is not necessary to do it twice. When in raw repl without soft reset already, we may stay there.

> Also, it doesn't really need to exit the raw repl.

I think it is necessary because the if I want to use these commands multiple times in single command chain there is no other way to do it, I must exit the raw repl and enter it again, e.g.

'mpremote **exec** "aaa=42" no-soft-reset _exec_ "aaa=aaa/7" _eval_ "aaa" soft-reset **exec** "pass"`

It will perform the soft reset before **bold** commands and no soft reset before _italic_ commands.

dpgeorge commented on tools/mpremote/mpremote/main.py:
```diff
@@ -463,9 +468,21 @@ def main():
+            if cmd == "soft-reset" and not soft_reset:
+                if pyb.in_raw_repl:
+                    pyb.exit_raw_repl()
+                soft_reset = True
```
I think the command name `soft-reset` is a bit confusing because it doesn't actually do a soft-reset.  Maybe `use-soft-reset` would be a better name.

Or actually make it do a soft reset, that might be better, like this:
```python
if pyb.in_raw_repl:
    pyb.exit_raw_repl()
pyb.enter_raw_repl()
```
Note that it wont do `soft_reset = True`, that would make the semantics hard to describe.

Or, for simplicity, just remove this command altogether and have only the `no-soft-reset` command (which should be enough for at least your original use case?).

jahr replied:
Having **no-soft-reset** comamnd only makes no sense in leaving it as a command because in such a case it cannot be used elsewhere than prior to first exec/eval/run command (otherwise the soft reset is already performed and it is performed only once on `enter_raw_repl()` in each comamnd chain), so I see two meaningful alternatives:

1. using `--no-soft-reset` command line option, which changes the default behavior of the command chain, i.e. no soft reset is performed on first exec/eval/run (my first commit) or
2. using commands `no-soft-reset` and `soft-reset` (renaming it to `with-soft-reset`, it' a good point). This will allow to control the soft reset behavior in more flexible way.

I would prefer the first one because it fulfills my use case (possibility to make multiple calls of exec/eval/run without soft resets). The second alternative probably makes no sense as I can't imagine any meaningful usecase for it.
I'm also thinking about another possibility:

3. Allow setting of soft_reset behavior be set in either config file or enviroment variable and allow users to change it with options or commands.

Let somone decide, I personally prefer alternative 1 ;-)

dpgeorge commented:
I made a PR for an alternative to this approach in #8347.

dpgeorge commented:
Thanks @jahr for opening this PR and getting the ball rolling on this issue.  It was a tough issue because of the subtleties.
